### PR TITLE
Support for sending the solution of the M8P RTK back to the GCS via Mavlink

### PR DESCRIPTION
**Pull request by rrr6399**
Closed this pull request since it was based on a branch with tow much other history in it: https://github.com/ArduPilot/ardupilot/pull/8564

### CODE REVIEW

File: libraries/AP_GPS/AP_GPS_ERB.cpp [hunk 1]
```diff
@@ -182,9 +182,10 @@ AP_GPS_ERB::_parse_gps(void)
             state.time_week_ms    = _buffer.stat.time;
             state.time_week       = _buffer.stat.week;
         }
+        Debug("Message ERB STAT num_sats= %u", _buffer.stat.satellites);
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_ERB.cpp]: The changes in this file seam to be unintended

Comment (rrr6399) [reply libraries/AP_GPS/AP_GPS_ERB.cpp]: I had put that in there when I was testing the Reach as well when they weren't sending the hdop and satellites in non-RTK mode. I can remove it.

**Comment by rrr6399** [comment]
@amilcarlucas Looks like the build failed due to a process error. Can that be restarted?
(If not, I'll add a space a recommit the file.)

**Comment by amilcarlucas** [comment]
Please squash the first with the third commit.

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 2]
```diff
@@ -851,6 +865,60 @@ AP_GPS_UBLOX::_parse_gps(void)
         state.horizontal_accuracy = 0;
 #endif
         break;
+
+    case MSG_NAV_RELPOSNED: {
+        Debug("MSG_NAV_RELPOSNED RTK status Received");
+
+        state.rtk_age_ms    = 0xFFFFFFFF;
+
+    	state.rtk_baseline_x_mm = _buffer.rtk.rel_pos_n_cm*10+_buffer.rtk.rel_pos_hp_n_mm/10.0;
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_UBLOX.cpp]: Remove the `.0` you don't want to be promoting to float here.

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 3]
```diff
@@ -851,6 +865,60 @@ AP_GPS_UBLOX::_parse_gps(void)
...
+    	state.rtk_baseline_y_mm = _buffer.rtk.rel_pos_e_cm*10+_buffer.rtk.rel_pos_hp_e_mm/10.0;
+    	state.rtk_baseline_z_mm = _buffer.rtk.rel_pos_d_cm*10+_buffer.rtk.rel_pos_hp_d_mm/10.0;
+
+    	int32_t acc_n_mm = _buffer.rtk.acc_n_mm;
+    	int32_t acc_e_mm = _buffer.rtk.acc_e_mm;
+    	int32_t acc_d_mm = _buffer.rtk.acc_d_mm;
+
+    	// rms of error terms
+    	double rtk_accuracy = acc_n_mm*acc_n_mm + acc_e_mm*acc_e_mm + acc_d_mm*acc_d_mm;
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_UBLOX.cpp]: There's some serious type mismatching going on here, you are doing all your math as int's and then storing into a double, but you've already risked overflowing in the pure integer math before converting to a double. You then pass this to `safe_sqrt()` which is going to give you a float result, so you might as well have stored this in a float in the first place, particularly because this is just a monitor, and you shouldn't ever appreciate the resolution difference here, since this is going to be dominated by casting into a unsigned type.

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 4]
```diff
@@ -851,6 +865,60 @@ AP_GPS_UBLOX::_parse_gps(void)
...
+    case MSG_NAV_RELPOSNED: {
+        Debug("MSG_NAV_RELPOSNED RTK status Received");
+
+        state.rtk_age_ms    = 0xFFFFFFFF;
+
+    	state.rtk_baseline_x_mm = _buffer.rtk.rel_pos_n_cm*10+_buffer.rtk.rel_pos_hp_n_mm/10.0;
+    	state.rtk_baseline_y_mm = _buffer.rtk.rel_pos_e_cm*10+_buffer.rtk.rel_pos_hp_e_mm/10.0;
+    	state.rtk_baseline_z_mm = _buffer.rtk.rel_pos_d_cm*10+_buffer.rtk.rel_pos_hp_d_mm/10.0;
+
+    	int32_t acc_n_mm = _buffer.rtk.acc_n_mm;
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_UBLOX.cpp]: Probably should be tagging these as `const`.

File: libraries/AP_GPS/AP_GPS_UBLOX.h [hunk 5]
```diff
@@ -101,6 +103,8 @@ class AP_GPS_UBLOX : public AP_GPS_Backend
 
     static bool _detect(struct UBLOX_detect_state &state, uint8_t data);
 
+    bool supports_mavlink_gps_rtk_message() { return true; }
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_UBLOX.h]: This should be made conditional on actually having received some none 0 information into the state. It's otherwise quite a waste of MAVLink bandwidth for most users that unless you are receiving the packet from the GPS (and potentially only if receiving it + actually in a RTK fix state) should it be sent.

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp (hunk 3 above)
Comment (WickedShell) [reply libraries/AP_GPS/AP_GPS_UBLOX.cpp]: Float won't overflow, and won't lose precision unless you have massive values. I'd prefer this to be float, because on most devices you just forced it into a very slow software path to do the double calculation. Since it's just a monitoring, and on something we shouldn't notice the loss on anyways) I'd like to see this as float.

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 6]
```diff
@@ -851,6 +865,60 @@ AP_GPS_UBLOX::_parse_gps(void)
         state.horizontal_accuracy = 0;
 #endif
         break;
+
+    case MSG_NAV_RELPOSNED: {
+        Debug("MSG_NAV_RELPOSNED RTK status Received");
+
+        state.rtk_age_ms    = 0xFFFFFFFF;
```
[the remark this answers is not in the document]
Comment (WickedShell) [reply libraries/AP_GPS/AP_GPS_UBLOX.cpp]: I was thinking about that. I think what you have is the bare minimum, and at least it's somewhat accurate. You will find that a GPS won't always use the most recent injected data (either not enough, waiting for more, or some internal thing I don't know of), so faking it gives us insight into if we are getting data up, but doesn't actually offer any insight into if the GPS is using the data, which is what the field is meant for.

Basically I'm not sure, and I don't feel strongly about it either way :)

**Comment by WickedShell** [comment]
> @WickedShell, which part of the patch was the comment referring to? What is your recommended solution?

Asking for the message from the GPS at all. The only solutions I have are to:

1. Change the desired default baud rate to be something faster that could accommodate all the messages. (This is probably a bit more work, as it requires some more testing, but fixes the larger issue, and reduces helps reduce transport delay of all GPS data)
2. Change it to not request the message if GPS_RAW_DATA is enabled. (Although this then means that you might want raw data + RTK and didn't get both)

**Comment by rrr6399** [comment]
Can we move forward with this since the message is being requested now in ArduCopter 4.0?

> > @WickedShell, which part of the patch was the comment referring to? What is your recommended solution?
> 
> Asking for the message from the GPS at all. The only solutions I have are to:
> 
> 1. Change the desired default baud rate to be something faster that could accommodate all the messages. (This is probably a bit more work, as it requires some more testing, but fixes the larger issue, and reduces helps reduce transport delay of all GPS data)
> 2. Change it to not request the message if GPS_RAW_DATA is enabled. (Although this then means that you might want raw data + RTK and didn't get both)

**Comment by WickedShell** [comment]
@rrr6399 Somewhat. It's not actually requested, we just happen to parse it if it's sent to us. The intended use case was also using it with manually configured GPS's. The pending PR for doing automatic moving baseline does turn it on, but it only turns it on if you are the moving base rover, and only works if it's an F9. That PR is also boosting the baud rate on the GPS. With that being said I'm advocating that we boost the baud rate in the general case, which would fix the problem here with baud rates. The only other one (which I only just caught) is to restrict the request for RELPOSNED to be hardware that is >= 8, but we're definitely closer if you want to try and resolve those aspects.

**Comment by rrr6399** [comment]
@WickedShell 

I hadn't seen any timing issues with requesting the RELPOSNED messages. But, I'll look into increasing the baud rate. I'll add a check for >=8 as well. Right now the auto-config logic isn't working for my M8P. I have had to manually turn on the RELPOSNED message. I need to troubleshoot why that is.

Also, the buffer size of 720 bytes used to store the injected RTK observations is too small when you get a lot of satellites (I've seen up to 30 satellites for the F9P) and when you use MSM7 messages that u-blox recommended for the moving base scenario. 

1.) Could the segment size be increased from 4 to at least 8 to accommodate all the observations?

I have seen GPS is not healthy messages when the base station is configured for moving base and the M8P sends the u-blox 4072 RTCM3 PVT message. The copter's GPS uses the message, but I wonder if there is a delay more than 240 ms. 

2.) Have you seen any delay issues with using moving base mode on the F9P? 

3.) Are you all also testing the moving base scenario where the base is not on the copter? 

4.) Since the RELPOSNED message is now being processed in 4.0, could the MSG_GPS_RTK be populated and sent as well similar to what was done in this pull request? (This pull request will likely have many conflicts with the new 4.0 implementation.)

> @rrr6399 Somewhat. It's not actually requested, we just happen to parse it if it's sent to us. The intended use case was also using it with manually configured GPS's. The pending PR for doing automatic moving baseline does turn it on, but it only turns it on if you are the moving base rover, and only works if it's an F9. That PR is also boosting the baud rate on the GPS. With that being said I'm advocating that we boost the baud rate in the general case, which would fix the problem here with baud rates. The only other one (which I only just caught) is to restrict the request for RELPOSNED to be hardware that is >= 8, but we're definitely closer if you want to try and resolve those aspects.

**Comment by WickedShell** [comment]
The M8P does not meet the timing requirements ArduPilot has for moving baseline, don't bother chasing that unless you want to carry a bunch of hard coded patches. An M8P can only provide a 4Hz moving baseline position [reference Page 6, 1-3 Performance](https://www.u-blox.com/sites/default/files/NEO-M8P_DataSheet_%28UBX-15016656%29.pdf). ArduPilot requires faster then 4Hz GPS data (we want 5Hz, and add a bit of tolerance for parsing jitter), and anytime the GPS data is older then that the GPS becomes unhealthy and the EKF won't fuse it. I'd recommend swapping to F9P's and saving yourself the development time you will spend chasing this.

The segment count for injection is dictated by the insufficient number of bits in the header for [GPS_RTCM_DATA](https://mavlink.io/en/messages/common.html#GPS_RTCM_DATA). The whole message is terrible though, because it could have been 255 bytes of payload instead of the 182 it currently is. You can fracture your message across multiple GPS_RTCM_DATA messages, but it can lead to the driver stepping on data. Just to be clear though, are you sending all your correction data up in one GPS_RTCM_DATA packet, or are you sending each RTCM message in it's own GPS_RTCM_DATA packet? (The later is strongly recommend for the record as it helps avoid the overflow)

Regarding generally looking at delay, look into GPA.Delta in your dataflash logs, this is the time in milliseconds since we parsed a complete set of required position data (on a M8/F9 it's going to be driven by parsing the NAV_PVT). This number should be 200 ms or better (depending upon GPS update rate), and anything more then +/- 20 ms of jitter is a bad sign, particularly if it happens a lot.

**Comment by rrr6399** [comment]
@WickedShell Ouch! Thanks for the heads up on the timing. I'll run some experiments to see if there is anyway to get the rate to 200 ms/fix and ask one of the experts on the u-blox forum to see if there is anyway around the 4 Hz rate. I hadn't seen that spec. 

As for bundling the RTCM observations, I've been passing through the RTCM message of each type in its own MAVLink GPC_RTCM_DATA message.  But, the RTCM message for say all the GPS satellites can be larger than the allotted buffer space if MSM7 format is used. The u-blox document mandates MSM7 for the moving base mode. (33.2.4 "The moving baseline reference message (type 4072, sub-type 0) must be used in combination with MSM7 observation messages.") I would have to programatically split the GPS MSM7 message into multiple MSM7 messages to be able to send it without increasing the destination buffer size and then I don't know what the implications are for processing it on the receiver side like you mentioned.

**Comment by WickedShell** [comment]
@rrr6399 You may be able to get it, but I suspect it will preform similarly to what happens if you ask for decent rate update intervals on more constellations then is listed. It may emit data with roughly that frequency, but the timing is terribly irregular, and some of the data is clearly wrong or extrapolated.

**Comment by rrr6399** [comment]
I looked at the delta time in the GPA message and found that the moving base solution takes between 500 and 1000 ms, but if I send a 1005 RTCM message and then switch back to the 4072 PVT message that the GPS continues to send fixes every 200 ms. I'm looking into whether the GPS is remaining in static base mode even though it received a 4072 message or if somehow switching modes fixed the timing issue.  
delta-time-rtk-moving-static-moving

**Comment by amilcarlucas** [comment]
If I remember correctly RTK needs a single message of type 1005, 1006, or 1007 to work.
Those need to be seen at least once.

**Comment by rrr6399** [comment]
It appears that once a  RTCM3 1005 message is received by the Rover GPS, it will always be in static baseline mode. Sending the 4072 PVT message after stopping the 1005 messages would not put the Rover into moving baseline mode. (I verified the mode by looking at the moving baseline flag of the NAV RelPosNed message in u-center.)

One advantage in this mode (initial static baseline with a switch over to moving baseline mode) is that I can get live PVTs from the base station for display purposes and use the reference position of the base for RTK processing to get accurate relative positions similar to the Reach. I'm going to do some tests in the field to see if this is a viable approach.

I also tried using a direct USB connection to the Pixhawk at 115 kbs rather than radio telemetry at 56 kbs to see if that helped the fix rate, but it didn't. The fix rate in moving baseline mode still remained between 400 to 1000 ms per fix rather than the required 200 ms. I attached a plot of the delta times per fix below when the rover was in moving baseline mode:

moving-base-delta-time

**Comment by amilcarlucas** [comment]
Why does this include chibiOS changes? I do not see where mavlink messages get sent.

**Comment by khancyr** [comment]
@WickedShell @rrr6399 hello guys, do you still want to pursuit this PR ? it is pretty old and would need huge rework.

**Comment by rrr6399** [comment]
I can submit a pull request for ArduCopter 4.0 and master. I have it working for 4.0.7 right now in my own branch.

> Upps, I missed that this is for ArduCopter3.6 branch, not for master, hence no conflicts ... :(

**Comment by amilcarlucas** [comment]
@rrr6399 can you rebase on 4.1 and close this one?

**Comment by peterbarker** [comment]
Thanks for this effort.  ArduPilot does consume this message now, so I think we've done what this PR set out to accomplish.  

Please reopen if I'm wrong.
